### PR TITLE
Add basic support for Tiny C Compiler

### PR DESCRIPTION
**Pull request by ret2libc**
This PR tries to add basic support tor tcc, Tiny C Compiler https://bellard.org/tcc/ .
I was able to use this to compile [Rizin](https://rizin.re) with tcc, however it probably still needs some adjustments because some unit tests still fail.

These are the failing tests: https://gist.github.com/ret2libc/1f386c7230e17a2c31be9ebf2c359523

Any hint on how to improve this support is welcome.

Do you suggest me to add `tcc` to the `mesonbuild/eoan` container and improve `os_comp.yaml` to test it?

### CODE REVIEW

**Comment by ret2libc** [comment]
`run_unittests.py::LinuxlikeTests::test_usage_pkgconfig_prefixes`: the TCC code does not add the rights `-L` necessary for TCC to find the libval1.so when looking at libval2.so

**Comment by ret2libc** [comment]
With newer version, these are the remaining failing tests:
```
FAILED run_unittests.py::AllPlatformTests::test_custom_target_changes_cause_rebuild - AssertionError: Lists differ: [] != ['prog']
FAILED run_unittests.py::AllPlatformTests::test_source_changes_cause_rebuild - AssertionError: Lists differ: [] != ['prog']
FAILED run_unittests.py::AllPlatformTests::test_nostdlib - subprocess.CalledProcessError: Command '['/usr/bin/ninja', '-w', 'dupbuild=err', '-d', 'explain']' returned non-zero exit status 1.
FAILED run_unittests.py::AllPlatformTests::test_preprocessor_checks_CPPFLAGS - subprocess.CalledProcessError: Command '['/home/rschiron/projects/meson/venv/bin/python3', '/home/rschiron/p...
FAILED run_unittests.py::LinuxlikeTests::test_compiler_check_flags_order - subprocess.CalledProcessError: Command '['/home/rschiron/projects/meson/venv/bin/python3', '/home/rschiron/proje...
FAILED run_unittests.py::LinuxlikeTests::test_static_link - subprocess.CalledProcessError: Command '['/usr/bin/ninja', '-w', 'dupbuild=err', '-d', 'explain']' returned non-zero exit statu...
FAILED run_unittests.py::LinuxlikeTests::test_usage_pkgconfig_prefixes - subprocess.CalledProcessError: Command '['/usr/bin/ninja', '-w', 'dupbuild=err', '-d', 'explain']' returned non-ze...
```

With the following patch I'm also able to make `test_static_link` pass, however I am not sure whether it is the right fix or not. Tcc does not have `-Wl,--start-group`/`--end-group`, so it is important to put dependencies in order, or so it seems. I would wait a bit before committing this fix to the test because maybe it is caused by other issues in the TCC plugin I wrote:
```diff
diff --git a/test cases/unit/68 static link/meson.build b/test cases/unit/68 static link/meson.build
index dac17f810..aff13cd2b 100644
--- a/test cases/unit/68 static link/meson.build        
+++ b/test cases/unit/68 static link/meson.build        
@@ -22,10 +22,10 @@ test('test3-shared', executable('test3-shared', 'test3.c',
 func7_dep = cc.find_library('func7')
 func9_linkwhole_dep = cc.find_library('func9_linkwhole')
 test('test4-linkwhole', executable('test4-linkwhole', 'test4.c',
-  dependencies : [func7_dep, func9_linkwhole_dep]))
+  dependencies : [func9_linkwhole_dep, func7_dep]))
 func9_linkwith_dep = cc.find_library('func9_linkwith')
 test('test4-linkwith', executable('test4-linkwith', 'test4.c',
-  dependencies : [func7_dep, func9_linkwith_dep]))
+  dependencies : [func9_linkwith_dep, func7_dep]))
 
 # Verify that installed libfunc16.a is usable
 libfunc16_dep = cc.find_library('func16')
```

**Comment by ret2libc** [comment]
By the way, to install Tiny C Compiler you can do:
```
          git clone https://repo.or.cz/tinycc.git
          cd tinycc
          git checkout mob
          chmod +x ./configure
          ./configure --prefix=/usr
          make
          sudo make install
```

**Comment by dcbaker** [comment]
This looks pretty good, I can spot a few little style things, but nothing looks majorly wrong. It would be nice if we a test run with tcc, but I think the most likely place to put one is in Eoan and that's broken if you try to rebuild it, and I'm going to fix that tomorrow.

**Comment by ret2libc** [comment]
> This looks pretty good, I can spot a few little style things, but nothing looks majorly wrong. It would be nice if we a test run with tcc, but I think the most likely place to put one is in Eoan and that's broken if you try to rebuild it, and I'm going to fix that tomorrow.

I'm still failing to understand why tcc is not re-called after touching a header file in test_source_changes_cause_rebuild

**Comment by jpakkane** [comment]
That requires the compiler to support Make style dependency output generation similar to GCC's `-MD` and friends. So either those flags are missing or they get written to the wrong place or something similar.

**Comment by dcbaker** [comment]
You need to implement the `get_dependency_gen_args` method, and maybe the `get_depfile_suffix` method

**Comment by ghostmansd** [comment]
What a funny coincidence. I should have checked this before starting working on the same idea (https://github.com/ghostmansd/meson/tree/tcc). :-) Thanks for your work! Your version seems to be much more complete.

File: mesonbuild/compilers/c.py [hunk 1]
```diff
@@ -710,3 +710,66 @@ def get_include_args(self, path: str, is_system: bool) -> T.List[str]:
...
+        return args
+
+    def get_optimization_args(self, optimization_level: str) -> T.List[str]:
+        return []
+
+    def get_no_optimization_args(self) -> T.List[str]:
+        return []
+
+    def get_pic_args(self) -> T.List[str]:
+        return []
```
Comment (ghostmansd) [inline mesonbuild/compilers/c.py:752]: It seems that -fPIC is not supported. Maybe it's worth throwing a MesonException here?

File: mesonbuild/linkers.py [hunk 2]
```diff
@@ -1372,3 +1372,88 @@ def get_soname_args(self, env: 'Environment', prefix: str, shlib_name: str,
...
+        return self._apply_prefix('--whole-archive') + args + self._apply_prefix('--no-whole-archive')
+
+    def get_coverage_args(self) -> T.List[str]:
+        return []
+
+    def export_dynamic_args(self, env: 'Environment') -> T.List[str]:
+        m = env.machines[self.for_machine]
+        if m.is_windows() or m.is_cygwin():
+            return self._apply_prefix('--export-all-symbols')
+        return self._apply_prefix('-export-dynamic')
```
Comment (ghostmansd) [inline mesonbuild/linkers.py:1398]: It seems that the tcc's equivalent of "-Wl,--export-dynamic" is "-rdynamic" option.

https://lists.gnu.org/archive/html/libtool-patches/2014-07/msg00001.html

File: mesonbuild/compilers/c.py [hunk 3]
```diff
@@ -710,3 +710,66 @@ def get_include_args(self, path: str, is_system: bool) -> T.List[str]:
...
+        return opts
+
+    def get_always_args(self) -> T.List[str]:
+        return []
+
+    def get_option_compile_args(self, options: 'KeyedOptionDictType') -> T.List[str]:
+        args = []
+        std = options[OptionKey('std', lang=self.language, machine=self.for_machine)]
+        if std.value != 'none':
+            args.append('-std=' + std.value)
```
Comment (ghostmansd) [inline mesonbuild/compilers/c.py:742]: I've been under impression that -std option is simply ignored with tcc. Please, correct me if I'm wrong.

File: mesonbuild/compilers/c.py [hunk 4]
```diff
@@ -710,3 +710,66 @@ def get_include_args(self, path: str, is_system: bool) -> T.List[str]:
...
+        return args
+
+    def get_optimization_args(self, optimization_level: str) -> T.List[str]:
+        return []
```
Comment (ghostmansd) [inline mesonbuild/compilers/c.py:746]: It looks like that tcc supports -Ox notation, but I'm not sure whether it has some effects other than defining an __OPTIMIZE__ macro.

File: mesonbuild/compilers/c.py [hunk 5]
```diff
@@ -710,3 +710,66 @@ def get_include_args(self, path: str, is_system: bool) -> T.List[str]:
...
+
+    def get_coverage_args(self) -> T.List[str]:
+        return []
+
+    def get_warn_args(self, level: str) -> T.List[str]:
+        warn_args = {
+            '0': [],
+            '1': ['-Wall'],
+            '2': ['-Wall', '-Werror'],
+            '3': ['-Wall', '-Werror', '-Wunsupported', '-Wwrite-strings']
```
Comment (ghostmansd) [inline mesonbuild/compilers/c.py:762]: Maybe '-Wimplicit-function-declaration' should also be here?

Comment (ghostmansd) [reply mesonbuild/compilers/c.py:762]: Never mind, it seems that one is included by -Wall already.

File: mesonbuild/linkers.py (hunk 2 above)
Comment (ghostmansd) [reply mesonbuild/linkers.py:1398]: With tcc 0.9.27, it seems that '-Wl,--export-all-symbols' is the exact equivalent of '-rdynamic'.

File: mesonbuild/compilers/c.py [hunk 6]
```diff
@@ -710,3 +710,66 @@ def get_include_args(self, path: str, is_system: bool) -> T.List[str]:
         if path == '':
             path = '.'
         return ['--include_path=' + path]
+
+class TCCCCompiler(CCompiler):
```
Comment (ghostmansd) [inline mesonbuild/compilers/c.py:714]: It looks like tcc has some support for -MD and -MF options; if I'm not mistaken these can be set at get_dependency_gen_args.

File: mesonbuild/linkers.py [hunk 7]
```diff
@@ -1372,3 +1372,88 @@ def get_soname_args(self, env: 'Environment', prefix: str, shlib_name: str,
...
+
+    def build_rpath_args(self, env: 'Environment', build_dir: str, from_dir: str,
```
Comment (ghostmansd) [inline mesonbuild/linkers.py:1400]: This section should likely be re-considered; I'm not sure of SunOS or cygwin parts, quite likely the logic can be simplified.

File: mesonbuild/linkers.py [hunk 8]
```diff
@@ -1372,3 +1372,88 @@ def get_soname_args(self, env: 'Environment', prefix: str, shlib_name: str,
...
+    def __init__(self, for_machine: mesonlib.MachineChoice,
+                 *, version: str = 'unknown version'):
+        super().__init__(['tcc'], for_machine, '-Wl,', [],
+                         version=version)
+
+    def get_allow_undefined_args(self) -> T.List[str]:
+        return []
+
+    def get_pic_args(self) -> T.List[str]:
+        return []
```
Comment (ghostmansd) [inline mesonbuild/linkers.py:1384]: cf. previous comment w.r.t. -fPIC

**Comment by eli-schwartz** [comment]
> It would be nice if we a test run with tcc, but I think the most likely place to put one is in Eoan and that's broken if you try to rebuild it, and I'm going to fix that tomorrow.

We don't need Ubuntu Eoan to run tcc unittests, it is packaged on Arch too.

File: mesonbuild/compilers/c.py (hunk 4 above)
Comment (ret2libc) [reply mesonbuild/compilers/c.py:746]: https://man.cx/tcc(1) here it says that -Ox options are ignored. That's why I did not list them here.

File: mesonbuild/compilers/c.py (hunk 3 above)
Comment (ret2libc) [reply mesonbuild/compilers/c.py:742]: I'm not sure...
```
$ tcc --help | grep std
  -std=c99     Conform to the ISO 1999 C standard (default).
  -std=c11     Conform to the ISO 2011 C standard.
```

File: mesonbuild/linkers.py (hunk 2 above)
Comment (ret2libc) [reply mesonbuild/linkers.py:1398]: From man page:
```
-Wl,--export-all-symbols
-Wl,--export-dynamic
           Export global symbols to the dynamic linker. It is useful when a library opened with "dlopen()" needs to access executable symbols.

```
I'm not sure there is really a need to special case windows/cygwin, I probably copied that code from another linker around :)

File: mesonbuild/compilers/c.py (hunk 6 above)
Comment (ret2libc) [reply mesonbuild/compilers/c.py:714]: I am looking into this, because that is causing some unit tests to fail, however it is not working as I would expect yet.

File: mesonbuild/compilers/c.py (hunk 3 above)
Comment (ghostmansd) [reply mesonbuild/compilers/c.py:742]: I checked the code; it seems that the only thing it affects is __STDC_VERSION__. This, however, seems to be an already sufficient reason to support it. :-)

File: mesonbuild/compilers/c.py (hunk 4 above)
Comment (ghostmansd) [reply mesonbuild/compilers/c.py:746]: This is used to define `__OPTIMIZE__` only.

File: mesonbuild/linkers.py (hunk 2 above)
Comment (ghostmansd) [reply mesonbuild/linkers.py:1398]: > I probably copied that code from another linker around
That's exactly the way I was doing the same thing. :-)

File: mesonbuild/compilers/c.py (hunk 4 above)
Comment (ret2libc) [reply mesonbuild/compilers/c.py:746]: so you think it makes sense to list the various `-Ox` here? If so I'll add.

File: mesonbuild/compilers/c.py (hunk 1 above)
Comment (ret2libc) [reply mesonbuild/compilers/c.py:752]: Raising an exception here makes *a lot* of unit tests fail. I think even tests that compile shared libraries fail if i put an exception.

File: mesonbuild/linkers.py (hunk 8 above)
Comment (ret2libc) [reply mesonbuild/linkers.py:1384]: exception here works fine, it seems.

File: mesonbuild/compilers/c.py (hunk 4 above)
Comment (ghostmansd) [reply mesonbuild/compilers/c.py:746]: Well, it's basically the same as for `-std`: the option only defines one macro, but still it works when passed. I think yes, we should be consistent. Perhaps `-Og` must be translated into `-O0 -g`, but I'm not sure.

File: mesonbuild/linkers.py (hunk 7 above)
Comment (ret2libc) [reply mesonbuild/linkers.py:1400]: I removed the SunOS part, but honestly i don't know how to improve this further without breaking stuff. I copied most of the body from gnu IIRC.

File: mesonbuild/compilers/c.py (hunk 6 above)
Comment (ret2libc) [reply mesonbuild/compilers/c.py:714]: Pushed a version with the -MD, -MF, however it causes other fails because of `assertBuildIsNoop`. Still not sure why.

**Comment by ret2libc** [comment]
> > It would be nice if we a test run with tcc, but I think the most likely place to put one is in Eoan and that's broken if you try to rebuild it, and I'm going to fix that tomorrow.
> 
> We don't need Ubuntu Eoan to run tcc unittests, it is packaged on Arch too.

Is there already a way to run unit tests for a particular compiler that does not involve Eoan? I just see a matrix for different compilers in os_comp.yml, but it is only on eoan. The `arch` job in os_comp.yml does not seem to allow different CC for now... Just wanted to know if there is already a way to do this, otherwise I will adjust some jobs.

**Comment by ret2libc** [comment]
> You need to implement the `get_dependency_gen_args` method, and maybe the `get_depfile_suffix` method

@dcbaker I have implemented the method in https://github.com/mesonbuild/meson/pull/8248/files#diff-7ada83df1272b6eaaff2fea8bbbb14ff9cd6baaf7de40c532fe95acf9bf73fbcR745, however now almost all tests that use `assertBuildIsNoop` fail because the test says stuff are recompiled when they should not. However, if I manually try to build those test directories and run ninja two times without touching any file, it doesn't recompile anything for me. Any idea where the problem might be?

**Comment by ret2libc** [comment]
> > > It would be nice if we a test run with tcc, but I think the most likely place to put one is in Eoan and that's broken if you try to rebuild it, and I'm going to fix that tomorrow.
> > 
> > 
> > We don't need Ubuntu Eoan to run tcc unittests, it is packaged on Arch too.
> 
> Is there already a way to run unit tests for a particular compiler that does not involve Eoan? I just see a matrix for different compilers in os_comp.yml, but it is only on eoan. The `arch` job in os_comp.yml does not seem to allow different CC for now... Just wanted to know if there is already a way to do this, otherwise I will adjust some jobs.

BTW, there is problem in using tcc from arch linux. That's the release of tcc, which is from about 3 years ago and it was not updated since then. There are bugs that are only fixed in the `mob` branch on https://repo.or.cz/tinycc.git, where the development seems to be right now. For example, when using `-pthread` the `main` function is not identified, and probably other issues. I think it's better to test with updated tcc from the `mob` branch if that needs to be done.

**Comment by dcbaker** [comment]
I bet @eli-schwartz can fix Arch ;)

**Comment by eli-schwartz** [comment]
https://repo.or.cz/h/mob.html

This is unstable development crowd-sourced by allowing fully anonymous push access to the world. The "mob" user carries no guarantees of usability, safety, or "doesn't contain malware".

So I rather suspect this would need to be carefully vetted before being packaged. :/ Does the tcc community / project leads have the intention of nominating some state in that branch as being known-good, able to graduate to the master branch, and maybe being tagged as a release (even a beta release)?

**Comment by ret2libc** [comment]
> https://repo.or.cz/h/mob.html
> 
> This is unstable development crowd-sourced by allowing fully anonymous push access to the world. The "mob" user/branch carries no guarantees of usability, safety, or "doesn't contain malware".
> 
> So I rather suspect this would need to be carefully vetted before being packaged. :/ Does the tcc community / project leads have the intention of nominating some state in that branch as being known-good, able to graduate to the master branch, and maybe being tagged as a release (even a beta release)?

I don't know :( I will send out an email on the mailing list, if there's anyone around, asking what are their intention.

**Comment by ret2libc** [comment]
Only failing tests now:
* AllPlatformTests::test_preprocessor_checks_CPPFLAGS
* AllPlatformTests::test_nostdlib
* LinuxlikeTests::test_compiler_check_flags_order
* LinuxlikeTests::test_usage_pkgconfig_prefixes
* LinuxlikeTests::test_static_link

File: mesonbuild/compilers/c.py [hunk 9]
```diff
@@ -710,3 +710,79 @@ def get_include_args(self, path: str, is_system: bool) -> T.List[str]:
...
```
Comment (dcbaker) [inline mesonbuild/compilers/c.py:714]: I know this is kinda pedantic, but can we call this `TinyCCmopiler`? Having 4 `C`s in a row is hard to read, and when you expand it makes no sense `TinyCCompilerCCompiler` :)

**Comment by dcbaker** [comment]
The compiler check order one is testing something that the `CCompileArgs` class does IIRC. You may need to implement one of those for TCC to pass that.

**Comment by XVilka** [comment]
@ret2libc was there any answer to your question in the TinyCC mailing list? Could you maybe invite their developers here, to this PR to take a look?

Also, could you please rebase it on top of the current master?

File: mesonbuild/compilers/c.py [hunk 10]
```diff
@@ -710,3 +710,79 @@ def get_include_args(self, path: str, is_system: bool) -> T.List[str]:
...
+    def get_optimization_args(self, optimization_level: str) -> T.List[str]:
+        optimization_args = {
+            '0': [],
+            'g': ['-O0', '-g'],
+            '1': ['-O1'],
+            '2': ['-O2'],
+            '3': ['-O3'],
+            's': ['-Os'],
+        }  # type: T.Dict[str, T.List[str]]
+        return optimization_args[optimization_level]
```
Comment (earboxer) [inline mesonbuild/compilers/c.py:757]: > Note: GCC options -Ox, -fx and -mx are ignored.

https://bellard.org/tcc/tcc-doc.html#Option-summary

Looking at the tinycc source, the optimizations flag is parsed (as one would expect according to `tcc -hh`), but it doesn't seem to be used at all (as one would infer from the documentation).

**Comment by earboxer** [comment]
> There are bugs that are only fixed in the mob branch on https://repo.or.cz/tinycc.git, where the development seems to be right now. For example, when using -pthread the main function is not identified, and probably other issues. I think it's better to test with updated tcc from the mob branch if that needs to be done.

> Does the tcc community / project leads have the intention of nominating some state in that branch as being known-good, able to graduate to the master branch, and maybe being tagged as a release (even a beta release)?

> was there any answer to your question in the TinyCC mailing list? 

There has not been a reply yet.

https://lists.nongnu.org/archive/html/tinycc-devel/2021-01/msg00221.html

tcc builds from the mob branch include the git hash in their version string with a prefix of "mob:"

```sh
./tcc -version # Latest mob branch
tcc version 0.9.27 mob:e97e108 (x86_64 Linux)
tcc -version # The official release from December 2017
tcc version 0.9.27 (x86_64 Linux)
```

This could be used to differentiate between versions: if it doesn't have the "mob" string in it, then it has XYZ known bugs/missing features.

But I think I agree with @TAAPArthur [who says](https://github.com/mesonbuild/meson/pull/8692#issuecomment-837377727)

> You know what you are getting into when dealing with development branches

It's nice to see some attempts toward https://github.com/mesonbuild/meson/issues/5406

**Comment by averms** [comment]
Ubuntu and Debian package a git snapshot from 2020-08-14. Nix (the unstable channel) packages a snapshot from 2021-10-09. The 2020-08-14 snapshot has `-MM` and `-MF` and the 2021-10-09 snapshot has `-MD` and `-MMD` too.

@ret2libc are you still working on this? I would love to help when I get some free time.

**Comment by ret2libc** [comment]
> Ubuntu and Debian package a git snapshot from 2020-08-14. Nix (the unstable channel) packages a snapshot from 2021-10-09. The 2020-08-14 snapshot has `-MM` and `-MF` and the 2021-10-09 snapshot has `-MD` and `-MMD` too.
> 
> @ret2libc are you still working on this? I would love to help when I get some free time.

I am focused on other things lately, so I don't have much time for moving this forward in the near future. AFAIR it should not require too much work to complete this PR though.

**Comment by dcbaker** [comment]
There's a few things to rebase, the compiler and linker detection has moved out of environment and into compilers/detect.py and linkers/detect.py, respectively. That looks scary, but should actually be pretty simple, JFYI

**Comment by ret2libc** [comment]
@averms did you already start working on this? Asking because I'm now having troubles with this branch because I need some newer feature and since I would have to rebase anyway I guess it would make sense to move this forward, but I'd like to avoid double work.

**Comment by averms** [comment]
> @averms did you already start working on this? Asking because I'm now having troubles with this branch because I need some newer feature and since I would have to rebase anyway I guess it would make sense to move this forward, but I'd like to avoid double work.

No I have not.

**Comment by ret2libc** [comment]
I'm fixing/skipping some tests... However, some of these tests require the `mob` version of tcc (which could change every day...). I'd like to know whether support for such a compiler would be accepted or not. In particular, I'm not sure I can add tests in CI for this, because arch has the 5-years old tcc.

Please help me to find the way forward.

**Comment by XVilka** [comment]
So what was decided? It would be nice to have that mainline, after all everyone who wants to use TCC knows what they are dealing with.

**Comment by ret2libc** [comment]
> So what was decided? It would be nice to have that mainline, after all everyone who wants to use TCC knows what they are dealing with.

Problem being that it is hard to even do some CI testing for it, because no distro ships it in the repos and you'd have to track a git repository that can change/break at any time, accessible by anyone.

**Comment by eli-schwartz** [comment]
> While tcc commits at the mob branch do get reverted occasionally (by the maintainer),

Indicating that the maintainer is still active; indeed, I notice that the person who tagged the last release has committed to `mob` as recent as 2 months ago (no real activity this month anyway).

That being the case, I'm extremely curious to know -- what are the chances of the maintainer(s) being willing to periodically "bless" the mob branch by pushing known-good revisions to master?

I mentioned this above:

> So I rather suspect this would need to be carefully vetted before being packaged. :/ Does the tcc community / project leads have the intention of nominating some state in that branch as being known-good, able to graduate to the master branch, and maybe being tagged as a release (even a beta release)?

**Comment by avih** [comment]
> what are the chances of the maintainer(s) being willing to periodically "bless" the mob branch by pushing known-good revisions to master?

TL;DR: impossible to tell.

This is basically a release. The last one was was some years ago, and mob had many commits since. It's been reqested to make a new release few times in recent years.

I also think soon would be a good time for a new release, but I can't speak for the maintainer.

If you want a "strategy" to choose a "safe" commit, I'd look for a commit by the maintainer (grischka), possibly with few further fixup commits, if there are any (he does tend to touch a lot of things in his commits, so sometimes it needs some followups).

**Comment by XVilka** [comment]
For everyone caring about TinyCC (`tcc`) support (https://github.com/mesonbuild/meson/pull/8248), they seem to be closer to start crafting releases and recently bumped the version: https://github.com/TinyCC/tinycc/commit/086870addd9599ef74bed8e81ebb353da7836709

**Comment by saidwho12** [comment]
Add support for TCC please this compiler is very fast
